### PR TITLE
analysis sub block

### PR DESCRIPTION
**Pull request by Superjomn**
The `subgraph_to_graph_pass` will create an IR graph for each operator having subblock.

This PR includes some changes to `graph_viz_pass` to draw the sub-graph.

TODO support IR analysis on sub-graphs.

# Demo
## rnn1
image

## dinger
dinge

### CODE REVIEW

File: paddle/fluid/framework/ir/subblock_to_graph_pass.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+#pragma once
+#include "paddle/fluid/framework/ir/pass.h"
+
+namespace paddle {
+namespace framework {
+namespace ir {
+
+// sub_block attribute name in OpDesc. TODO(Superjomn) unify these.
+const char kSubblockAttrKeyStr[] = "sub_block";
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/subblock_to_graph_pass.h]: Can replace every places that use "sub_block"

File: paddle/fluid/framework/ir/subblock_to_graph_pass.cc [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  // Init sub graph
+  if (!graph->Has(kSubblockGraphAttr)) {
+    graph->Set(kSubblockGraphAttr, new subgraphs_t);
+  }
+
+  auto& sub_graphs = graph->Get<subgraphs_t>(kSubblockGraphAttr);
+
+  // Filter out the nodes that has sub-block
+  for (auto* node : graph->Nodes()) {
+    if (node->IsOp() && node->Op()->HasAttr(kSubblockAttrKeyStr)) {
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/subblock_to_graph_pass.cc]: `IsOp && Op()`

File: paddle/fluid/framework/ir/subblock_to_graph_pass.cc [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+                                proto::ProgramDesc* program_desc) {
+  *program_desc->add_blocks() = *block_desc->Proto();
+}
+
+std::unique_ptr<ir::Graph> SubblockToGraphPass::ApplyImpl(
+    std::unique_ptr<ir::Graph> graph) const {
+  // Init sub graph
+  if (!graph->Has(kSubblockGraphAttr)) {
+    graph->Set(kSubblockGraphAttr, new subgraphs_t);
+  }
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/subblock_to_graph_pass.cc:31]: else return or throw error?

File: paddle/fluid/framework/ir/subblock_to_graph_pass.cc [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+      for (auto op : sub_block_desc->Proto()->ops()) {
+        LOG(INFO) << op.type() << " in subblock";
+      }
+
+      framework::proto::ProgramDesc fake_proto;
+      TransformSubblockToProgram(sub_block_desc, &fake_proto);
+      framework::ProgramDesc fake_program_desc(fake_proto);
+      // Create a graph
+      sub_graphs[node] = std::unique_ptr<Graph>(new Graph(fake_program_desc));
+      LOG(INFO) << "get sub-graph size " << sub_graphs[node]->Nodes().size();
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/subblock_to_graph_pass.cc]: VLOG?

File: paddle/fluid/framework/ir/graph_viz_pass.cc [hunk 5]
```diff
@@ -82,14 +95,53 @@ std::unique_ptr<ir::Graph> GraphVizPass::ApplyImpl(
...
+  int num_sub_graph{0};
+  auto node_num = static_cast<int>(graph.Nodes().size());
+
   // Create nodes
-  for (const Node* n : graph->Nodes()) {
-    std::string node_id = FormatName(n) + "(" + std::to_string(n->id()) + ")";
+  for (const Node* n : graph.Nodes()) {
+    std::string node_id =
+        FormatName(n) + "(" + std::to_string(n->id() + node_id_offset) + ")";
     if (n->IsOp()) {
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph_viz_pass.cc]: `IsOp && Op`

File: paddle/fluid/framework/ir/subblock_to_graph_pass.cc (hunk 3 above)
Comment (Superjomn) [reply paddle/fluid/framework/ir/subblock_to_graph_pass.cc:31]: Just init here, create a subgraphs if not exists. No need to raise error.

Comment (panyx0718) [reply paddle/fluid/framework/ir/subblock_to_graph_pass.cc:31]: Are you creating the subgraphs if exists? (graph->Has returns true?)

File: paddle/fluid/framework/ir/subblock_to_graph_pass.cc (hunk 4 above)
Comment (panyx0718) [reply paddle/fluid/framework/ir/subblock_to_graph_pass.cc]: VLOG?

File: paddle/fluid/framework/ir/subblock_to_graph_pass.cc (hunk 3 above)
Comment (Superjomn) [reply paddle/fluid/framework/ir/subblock_to_graph_pass.cc:31]: done

https://github.com/PaddlePaddle/Paddle/pull/14936/files#diff-f4c0665c5674c253c8aa0179cf6022f4R33

**Review by panyx0718 — DISMISSED** [review]

File: paddle/fluid/framework/ir/subblock_to_graph_pass.cc (hunk 4 above)
Comment (Superjomn) [reply paddle/fluid/framework/ir/subblock_to_graph_pass.cc]: done

**Pull request closed without merging**